### PR TITLE
Canonical correlations

### PR DESCRIPTION
**Pull request by stevekogo**
Initial implementation of Canonical Correlations

### CODE REVIEW

**Comment by stevekogo** [comment]
@rdstern We need to discuss the what needs to be implemented in this dialogue plus the layout both for the dialogue and the sub-dialogue.

File: instat/dlgCanonicalCorrelationAnalysis.vb [hunk 1]
```diff
@@ -15,8 +15,88 @@
 ' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 Imports instat.Translations
 Public Class dlgCanonicalCorrelationAnalysis
+    Public bFirstLoad As Boolean = True
+    Dim clsModel As New ROperator
+    'Public Variables As New RSyntax
+    Public YVariables, XVariables, dataframe
```
Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: What are these variables for?

Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: Is this just a temporary thing before you know the function and parameters?

**Comment by stevekogo** [comment]
This is just temporary implementation but i will cleanup the code.

**Comment by dannyparsons** [comment]
@volloholic will also want to be part of this discussion I think.

File: instat/dlgCanonicalCorrelationAnalysis.vb [hunk 2]
```diff
@@ -15,8 +15,80 @@
...
+    Private Sub Options()
+        Cancor()
+    End Sub
+
+    Public Function XYariables(ByVal my_variable) As String
+        Dim Variables As New RSyntax
+        Variables.SetFunction(frmMain.clsRLink.strInstatDataObject & "$get_columns_from_data")
+        Variables.AddParameter("col_names", my_variable)
+        Variables.AddParameter("data_name", Chr(34) & ucrSelectorCCA.ucrAvailableDataFrames.cboAvailableDataFrames.SelectedItem & Chr(34))
+        Return (Variables.GetScript())
```
Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: You definitely shouldn't have another RSyntax in here. You should always be using RSyntax from ucrBase.

File: instat/dlgCanonicalCorrelationAnalysis.vb [hunk 3]
```diff
@@ -15,8 +15,80 @@
...
+    End Sub
+
+    Private Sub ucrbaseCCA_clickok(sender As Object, e As EventArgs) Handles ucrBaseCCA.ClickOk
+        'sdgcanonicalcorrelation.regoptions
+        Options()
+    End Sub
+
+    Public Sub ucrReceiverYvariables_SelectionChanged() Handles ucrReceiverYvariables.SelectionChanged
+        TestOKEnabled()
+        YVariables = XYariables(ucrReceiverYvariables.GetVariableNames())
```
Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: I don't quite understand why you are not just doing:
`ucrBaseCCA.clsRsyntax.AddParameter("X", ucrReceiverYvariables.GetVariables())`

File: instat/dlgCanonicalCorrelationAnalysis.vb [hunk 4]
```diff
@@ -15,8 +15,80 @@
...
+    End Sub
+
+    Public Sub ucrReceiverXvariables_SelectionChanged() Handles ucrReceiverXvariables.SelectionChanged
+        TestOKEnabled()
+        XVariables = XYariables(ucrReceiverXvariables.GetVariableNames())
```
Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: And the same here for the "X" parameter.

File: instat/dlgCanonicalCorrelationAnalysis.vb [hunk 5]
```diff
@@ -15,8 +15,80 @@
...
+        End If
+    End Sub
+
+    Private Sub ucrBaseCCA_ClickReset(sender As Object, e As EventArgs) Handles ucrBaseCCA.ClickReset
+        SetDefaults()
+    End Sub
+
+    Private Sub ucrbaseCCA_clickok(sender As Object, e As EventArgs) Handles ucrBaseCCA.ClickOk
+        'sdgcanonicalcorrelation.regoptions
+        Options()
```
Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: If you use the RSyntax from ucrBase then all this is done automatically.

File: instat/dlgCanonicalCorrelationAnalysis.vb (hunk 3 above)
Comment (stevekogo) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: Both X and Y should be data frames. That is why I created a function XYariables to create data frames.

Comment (dannyparsons) [inline instat/dlgCanonicalCorrelationAnalysis.vb]: If there is more than one variable in the receiver, GetVariables() returns a data frame.
We should add an option in the R method for if you want to return a data frame when only one variable is selected too.

**Comment by stevekogo** [comment]
@dannyparsons @bethanclarke I have done some fixes on CCA but there seems to be a problem(dialogue crushing) if the dimensions of x variables is different from the dimension of y variables. This is also the case when I try it running the same code in R. 

Can you please look at this? I am using cancor function from candisc package.

**Comment by rdstern** [comment]
Hi Steve,

Not sure quite what you mean by “dimensions are different”?  The x and y variables should be able to have a different number of variables.  But they need the same number of rows – which may not be the case when you issue the instructions to cpt.

The dialogue should check that.

Cheers

Roger

**Comment by stevekogo** [comment]
@dannyparsons I have discussed with @volloholic that I fix the conflicts in a new commit. This commit is to be kept open but is not going to be merged.

**Comment by dannyparsons** [comment]
Ok, sounds good.

**Pull request closed without merging**